### PR TITLE
Produce MS Coff by default when targetting windows

### PR DESCRIPTION
**Pull request by thewilsonator**
This rebases #12825 but doesn't attempt to rename phobos libraries and doesn't change the build infra where possible(and doesn't have a large diff in `link.d`). See if this fixes the CI issues.

### CODE REVIEW

File: changelog/fix18964.dd [hunk 1]
```diff
@@ -0,0 +1,5 @@
+`-m32` now produces MS Coff objects when tergatting windows
```
Comment (ibuclaw) [inline changelog/fix18964.dd:1]: Tergatting?

**Comment by Geod24** [comment]
As much as I agree we should go in this direction, @WalterBright 's opposition is with the very essence of this PR.

**Comment by thewilsonator** [comment]
#12825 had a misleading title, and I can understand, to a certain extent, any opposition based on misunderstandings from that PR. 

That being said, none of Walter's points are backed up by current reality:

>> [the Microsoft toolchain does not come with Windows.](https://github.com/dlang/dmd/pull/12825#issuecomment-876388826)
> [we ship lld (LLVM based linker) and minGW import libraries on windows. Lack of toolchain is not a problem.](https://github.com/dlang/dmd/pull/12825#issuecomment-876398213)

>> 1. Versions. Every customer has a different version of that linker. Some would work, some wouldn't. I couldn't just ship someone with a non-working MS-Link a linker that did work.
>>2. Bugs. Microsoft placed no priority whatsoever on fixing bugs. I was faced with being dead in the water for indeterminate times. Obviously, this was untenable.
>>3. The compiler didn't work out of the box. I had to tell customers they had to buy some MS dev system to get a linker.
>>4. When MS-Link didn't work, or crashed, etc., I always got blamed.

https://github.com/dlang/dmd/pull/12825#issuecomment-876016998

File: changelog/fix18964.dd (hunk 1 above)
Comment (thewilsonator) [reply changelog/fix18964.dd:1]: Whoops, will fix tomorrow.

**Comment by WalterBright** [comment]
Leaving it the way it is allows us to control the out-of-box experience. Relying on the Microsoft toolchain by default does not. Minimizing the friction for people wanting to try out D is critical.

**Comment by maxhaton** [comment]
> Leaving it the way it is allows us to control the out-of-box experience. Relying on the Microsoft toolchain by default does not. Minimizing the friction for people wanting to try out D is critical.

Isn't Nicholas's point that we don't depend on the Microsoft toolchain because we ship LLVM's linker and minGW?

**Comment by WalterBright** [comment]
If I recall correctly, much of the impetus for this came from vibe.d causing optlink to fail. This turned out to not be a problem with optlink, but was caused by replacing the system lib files meant for optlink (in OMF format) with lib files from Microsoft, which were in MS-COFF format. The solution was to simply run `coffimplib.exe` on the new system .lib files, which is the responsibility of our team, not the customer. `coffimplib` converts the files to OMF format.

**Comment by WalterBright** [comment]
Microsoft runtime libraries and the Microsoft linker are required to compile with the `-m32mscoff` switch. They are not part of the Windows operating system distribution and are not on the user's machine by default. Ideally,
```
dmd hello
```
should be the user's initial experience after unzipping the download. Not "linker not found" and "library not found" errors. One could argue this isn't a big deal, but in my experience, you'll lose half the customers right off the bat with that.

**Comment by WalterBright** [comment]
Recall the success Python had with "batteries included" (and consumer products in general - I remember when things didn't come with batteries.)

**Comment by adamdruppe** [comment]
dmd -m32mscoff has worked out of the box for some time now. The dmd distribution has included the batteries for.... a year or two.

**Comment by WalterBright** [comment]
Ok. I have dmd 2.092 installed on my machine. I wrote:
```
import std.stdio;
void main {
    writeln("hello world\n");
}
```
compiling:
```
dmd hello -m32mscoff
lld-link: error: undefined symbol: __imp__InterlockedIncrement@4
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\tidtable.c:402
>>>               libcmt.lib(tidtable.obj):(__initptd)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\mbctype.c:524
>>>               libcmt.lib(mbctype.obj):($LN14)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\mbctype.c:618
>>>               libcmt.lib(mbctype.obj):(__setmbcp)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\localref.c:39
>>>               libcmt.lib(localref.obj):(___addlocaleref)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\setlocal.c:463
>>>               libcmt.lib(setlocal.obj):($LN10)

lld-link: error: undefined symbol: __imp__InterlockedDecrement@4
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\tidtable.c:596
>>>               libcmt.lib(tidtable.obj):(__freefls@4)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\mbctype.c:511
>>>               libcmt.lib(mbctype.obj):($LN14)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\mbctype.c:612
>>>               libcmt.lib(mbctype.obj):(__setmbcp)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\mbctype.c:638
>>>               libcmt.lib(mbctype.obj):(__setmbcp)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\localref.c:79
>>>               libcmt.lib(localref.obj):(___removelocaleref)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\initctyp.c:195
>>>               libcmt.lib(initctyp.obj):(___init_ctype)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\initctyp.c:227
>>>               libcmt.lib(initctyp.obj):($error_cleanup$28334)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\setlocal.c:260
>>>               libcmt.lib(setlocal.obj):(__free_locale)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\setlocal.c:905
>>>               libcmt.lib(setlocal.obj):(__setlocale_get_all)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\setlocal.c:921
>>>               libcmt.lib(setlocal.obj):($LN26)
>>> referenced 5 more times

lld-link: error: undefined symbol: __imp__InterlockedExchange@8
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\setmode.c:149
>>>               libcmt.lib(setmode.obj):(__set_fmode)
>>> referenced by f:\dd\vctools\crt_bld\self_x86\crt\src\rand_s.c:94
>>>               libcmt.lib(rand_s.obj):(_rand_s)
Error: linker exited with status 1
```
There is no `f:\dd` directory on my system.
```
dmd hello
E:\dmd2.092\windows\bin>dmd hello

E:\dmd2.092\windows\bin>hello
hello world


E:\dmd2.092\windows\bin>
```

**Comment by WalterBright** [comment]
P.S. I was in the `e:\dmd2.092\windows\bin` directory.

**Comment by thewilsonator** [comment]
what is the linker command issued by dmd? because the libraries are right there in the install

**Comment by WalterBright** [comment]
> what is the linker command issued by dmd?

```
E:\dmd2.092\windows\bin\lld-link.exe /NOLOGO "hello.obj"   /DEFAULTLIB:phobos32mscoff /OPT:NOICF  /LIBPATH:"C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\lib" /LIBPATH:".\..\lib32mscoff\mingw" /SAFESEH:NO
```
The weird `f:\dd` path is another problem. No, it is not in any of my environment variables.

**Comment by adamdruppe** [comment]
it is kinda weird that it includes a MSVC path in your link. that doesn't happen on my out-of-the-box setup that just works.

**Comment by thewilsonator** [comment]
Do you have a bjorked VC installation? 

On a freshly installed clean windows box with no VS installation it should be picking up the `_InterlockedIncrement` function in `.\..\lib32mscoff\mingw\kernel32.lib`.

**Comment by WalterBright** [comment]
So I need to delete my VC installation in order to use dmd? It's hardly bjorked, it's the totally default install. Also, when I use it, I use a separate command prompt with the VC environment variables. The command prompt I was running did not set that environment.

I still don't know where it is getting things from, nor where that `f:\dd` comes from.

If you say "you shouldn't be using VC anymore, it's obsolete, it won't work with dmd" ok, but that sort of thing was exactly why I got optlink in the first place. Telling customers they had to change which Microsoft dev products they had to use was a total non-starter with them. It invariably produced a "no".

**Comment by adamdruppe** [comment]
On Thu, Sep 30, 2021 at 05:34:07PM -0700, Walter Bright wrote:
> So I need to delete my VC installation in order to use dmd?

Did you change your sc.ini?

And is this from the zip or from the dmd installer? If the installer
maybe it detected the VC and edited sci.ini when it perhaps shouldn't
have.

**Comment by thewilsonator** [comment]
> nor where that f:\dd comes from.

FWIW, that's probably a Microsoft build farm internal path, I've seen similar things with with Apple and error messages from developer tool crashing and from kernel panics from all systems. Unless you built them, the debug info for e.g. the symbols corresponding to the C runtime startup routines weren't build on your machine and won't make sense on your filesystem.

**Comment by WalterBright** [comment]
> Did you change your sc.ini?

No. I just unzipped it.
```
E:\dmd2.092\windows\bin>type sc.ini
[Version]
version=7.51 Build 020


; environment for both 32/64 bit
[Environment]
DFLAGS="-I%@P%\..\..\src\phobos" "-I%@P%\..\..\src\druntime\import"

; optlink only reads from the Environment section so we need this redundancy
; from the Environment32 section (bugzilla 11302)
LIB="%@P%\..\lib"


[Environment32]
LIB="%@P%\..\lib"
LINKCMD=%@P%\optlink.exe


[Environment64]
LIB=%@P%\..\lib64
; needed to avoid COMDAT folding (bugzilla 10664)
DFLAGS=%DFLAGS% -L/OPT:NOICF

; -----------------------------------------------------------------------------
[Environment32mscoff]
LIB=%@P%\..\lib32mscoff
; needed to avoid COMDAT folding (bugzilla 10664)
DFLAGS=%DFLAGS% -L/OPT:NOICF
```

**Comment by thewilsonator** [comment]
>The command prompt I was running did not set that environment.

did you try one that does set them? You're supposed to use that one for using VS tools and things that rely on it. Perhaps detection of non-VS envvar command lines don't work when you have VS installed?

**Comment by WalterBright** [comment]
```
06/13/2020  12:38 AM        33,119,228 dmd.2.092.1.windows.7z
```

**Comment by thewilsonator** [comment]
> did you try one that does set them? You're supposed to use that one for using VS tools and things that rely on it. Perhaps detection of non-VS envvar command lines don't work when you have VS installed?

To put that another way: 
It is not unreasonable to expect users who have VS installed to run commands related to programming things from a VS command line. Similarly we expect people who have not installed VS to run such commands from a console that does not have the VS env vars set. For users that have VS installed the default libraries used should be the ones built by MS, the minGW ones are a fallback for a reason.

However I suspect that the edge case of having a VS installation AND not using a VS console is not handled properly. i.e. it detects a VS installation (which is correct, there is one) but does not account for the fact the env vars (or whatever the underlying issue is) are not properly set.

**Comment by WalterBright** [comment]
> did you try one that does set them?

No.

> You're supposed to use that one for using VS tools and things that rely on it. Perhaps detection of non-VS envvar command lines don't work when you have VS installed?

It doesn't say that in the `README.TXT`. Please understand, I know that you are trying to help me, and I appreciate that. But that isn't the point. The point is our customers seeing those problems as their very first interaction with D, what are they going to do? They're going to delete it and move on to Go/Rust/C++.

```
dmd hello
```
Must work out of the box, or we will immediately lose 50% of the people who are willing to give us a chance.

**Comment by 12345swordy** [comment]
Let me intrude and point out that it would be better for walter bright to share screens with @thewilsonator so that you two can figure out what the issue is.

**Comment by WalterBright** [comment]
I've cut&pasted from my screen. Since it's all text mode, that should be sufficient.

**Comment by adamdruppe** [comment]
On Thu, Sep 30, 2021 at 06:11:37PM -0700, Walter Bright wrote:
> Must work out of the box, or we will immediately lose 50% of the people who are willing to give us a chance.

idk why it doesn't for you, but it does for me on all computers.

one that has never had visual studio or any sdk installed, it just
works. -m32, -m32mscoff, and -m64 all the same. It uses bundled lld-link
there.

one that currently has visual studio installed, it just works.
Interestingly on that box though it does use Microsoft's linker as
confirmed by dmd -v.


<snip>
function  D main
function  hello._d_cmain!().main
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\bin\HostX64\x86\link.exe /NOLOGO "hello.obj"   /DEFAULTLIB:phobos32mscoff  /LIBPATH:"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\lib\x86" legacy_stdio_definitions.lib /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\ucrt\x86" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x86"


I actually find that a little bit weird. I don't know why it would be
looking for that when I haven't modified sc.ini. And MS link.exe is NOT
in the path.

does dmd go looking for that itself?!?

**Comment by maxhaton** [comment]
> On Thu, Sep 30, 2021 at 06:11:37PM -0700, Walter Bright wrote: Must work out of the box, or we will immediately lose 50% of the people who are willing to give us a chance.
> idk why it doesn't for you, but it does for me on all computers. one that has never had visual studio or any sdk installed, it just works. -m32, -m32mscoff, and -m64 all the same. It uses bundled lld-link there. one that currently has visual studio installed, it just works. Interestingly on that box though it does use Microsoft's linker as confirmed by dmd -v. <snip> function D main function hello._d_cmain!().main C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\bin\HostX64\x86\link.exe /NOLOGO "hello.obj" /DEFAULTLIB:phobos32mscoff /LIBPATH:"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\lib\x86" legacy_stdio_definitions.lib /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\ucrt\x86" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x86" I actually find that a little bit weird. I don't know why it would be looking for that when I haven't modified sc.ini. And MS link.exe is NOT in the path. does dmd go looking for that itself?!?

```d
/**
     * retrieve path to the Microsoft linker executable
     * also modifies PATH environment variable if necessary to find conditionally loaded DLLs
     * Params:
     *   x64 = target architecture (x86 if false)
     * Returns:
     *   absolute path to link.exe, just "link.exe" if not found
     */
    const(char)* linkerPath(bool x64)
    {
        const(char)* addpath;
        if (auto p = getVCBinDir(x64, addpath))
        {
            OutBuffer cmdbuf;
            cmdbuf.writestring(p);
            cmdbuf.writestring(r"\link.exe");
```
apparently?

**Comment by rainers** [comment]
>  /LIBPATH:"C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\lib" /LIBPATH:".\..\lib32mscoff\mingw"

Please note that Walter is using VC 10 from VS 2010, which is at least 6 major versions behind the current MS toolset. dmd doesn't seem to find the matching Windows SDK (if it is installed), so the mingw libraries are used. This combination doesn't seem to work. I doubt this affects a lot of users, though.

Maybe less controversial: build 64-bit applications on 64-bit hosts by default, which would use the MS-COFF tools just as well.

**Comment by thewilsonator** [comment]
> Please note that Walter is using VC 10 from VS 2010, which is at least 6 major versions behind the current MS toolset

Out of interest, what versions of VS _do_ we support?

**Comment by rainers** [comment]
> Out of interest, what versions of VS _do_ we support?

We  have support for anything starting from VC9, i.e. VS2008, see vsoptions.d. I guess not all combinations with mingw work, just as you cannot combine the VS toolchain with any Windows SDK. I suspect Walter might be able to link from within the VC command line with the environment variables set correctly.

> edit, or for that matter does MS support?

https://docs.microsoft.com/en-us/visualstudio/productinfo/vs2017-servicingarchive-vs :
"These products follow the Microsoft Product Lifecycle Policy of 10 years (5 years of Mainstream Support and 5 years of Extended Support), starting with the date the major product version is released to the world (RTW)"

So VS2010 is no longer supported.

**Comment by 12345swordy** [comment]
> 
> 
>Since it's all text mode, that should be sufficient.

Not quite, there is quite a lot of information that @thewilsonator can pick up that you overlook. Take from my work experience: a picture is worth a thousand words.

**Comment by thewilsonator** [comment]
@WalterBright I think I have fixed that particular issue, could you please retest (I have no windows box to confirm this).

**Comment by WalterBright** [comment]
It looks all right but I'd like those fixes in a separate PR, because it is a separate thing from making -m32mscoff the default.

**Comment by thewilsonator** [comment]
Filed as https://issues.dlang.org/show_bug.cgi?id=22355

**Comment by thewilsonator** [comment]
https://github.com/dlang/dmd/pull/13125

**Comment by RazvanN7** [comment]
@thewilsonator @WalterBright is this good to go?

**Comment by WalterBright** [comment]
1. has anyone tried a compile/link speed test?
2. a large part of this (most?) is a refactoring, i.e. replacing target.mscoff with target.omfobj, and then finding the uses and flipping them about. This is the wrong way to go. Just set target.mscoff to true by default. Mixing in the refactoring opens the possibility of bugs introduced by the refactorings.
3. the existing code has a poor design where !target.mscoff means generate OMF. These sorts of things inevitably lead to hidden bugs, as !target.mscoff may in the future mistakenly enable target.elfobj. This refactoring, by just flipping things about, perpetuates this risk rather than fixing it. One can see in the coding changes there is zero improvement to the logic in the refactored code.

BTW, a lot of the motivation for this change seems to stem with optlink suddenly starting to fail with vibe.d after vibe.d changed import libraries. The fault there was the new import libraries had not been converted using the coffimplib.exe tool.

https://www.digitalmars.com/ctg/coffimplib.html

**Comment by dkorpel** [comment]
> BTW, a lot of the motivation for this change seems to stem with optlink suddenly starting to fail with vibe.d after vibe.d changed import libraries.

I also recall the symbol count in vibe.d exceeding optlink's maximum of 32768/65536 (something 16-bit)

**Comment by thewilsonator** [comment]
1. No, but optlink is abandonwere, frequently breaks, can't compile vibe and many more issues beside. Correct function is more important than speed.
2. it is written so that omfobj is in general the exception to the normal flow of things. Note that there are 15 uses of `target.omfobj` and 5 uses of `!target.omfobj`
3. yes we should probably migrate to an enum of object formats, that is for another PR.

**Comment by ibuclaw** [comment]
More tests have been turned off because optlink isn't suitable for linking D code. Why isn't this merged yet?

https://github.com/dlang/druntime/pull/3667

**Comment by RazvanN7** [comment]
ping @WalterBright

**Comment by WalterBright** [comment]
> More tests have been turned off because optlink isn't suitable for linking D code. Why isn't this merged yet?

Nobody has spent a few minutes with bisecting the unittest in the PR to see what triggered the problem.

As for speed, speed is the big reason why people use dmd. Should at a bare minimum *find out* what the speed difference is using MS-LINK.

As usual, I object to 1 line of semantic change and 50 lines of refactoring in the same PR. I've explained why repeatedly.

**Comment by adamdruppe** [comment]
On Tue, Jan 11, 2022 at 03:03:22PM -0800, Walter Bright wrote:
> As for speed, speed is the big reason why people use dmd. Should at a bare minimum *find out* what the speed difference is using MS-LINK.

one of my gui programs average compile+link time over 3 runs:

2.7s with optlink -m32
2.6s with microsoft link -m32mscoff or -m64
2.2s with the bundled lld-link with -m32mscoff
2.3s with the bundled lld-link with -m64

In a hello world, I got 1.1s with MS link, 1.0 with optlink, 0.9 with
lld-link.

If you want speed, the best plan is probably to just use the
lld-link which is already bundled with the compiler release. It
consistently performs better.

**Comment by WalterBright** [comment]
@adamdruppe thank you, this is good information. How large is the resulting binary?

**Comment by adamdruppe** [comment]
On Tue, Jan 11, 2022 at 04:09:18PM -0800, Walter Bright wrote:
> @adamdruppe thank you, this is good information. How large is the resulting binary?

for the gui test program:

1.5 mb optlink
1.6 mb 32 bit build with the other two linkers
2.1 mb 64 bit build


I expect the small difference in file size is more due to C library than
linker.

**Comment by WalterBright** [comment]
I agree it should be the library.

Can you try a larger executable? Like maybe turn on symbolic debug info? A 2Mb executable may be more about loading the files than doing linking work.

**Comment by 12345swordy** [comment]
> I agree it should be the library.
> 
> Can you try a larger executable? Like maybe turn on symbolic debug info? A 2Mb executable may be more about loading the files than doing linking work.

Is compiling dmd, druntime, and phobos a good test case here for @adamdruppe then?

**Comment by maxhaton** [comment]
Reading the whole dmd binary (7MB) and computing the sha256 digest of takes less than 40ms on my machine so I don't think 2MB is an unreasonable place to draw a conclusion from.

I would also like to note that gdc-10 is almost 2x faster at hello world than dmd on my machine, and ever so slightly faster at a std.regex example.

**Comment by ibuclaw** [comment]
> I would also like to note that gdc-10 is almost 2x faster at hello world than dmd on my machine, and ever so slightly faster at a std.regex example.

That could be linking to a dynamic library vs static though, which can make the difference between 40ms vs 400ms link times.

**Comment by maxhaton** [comment]
Sure, probably true, but my point is that dmd looks much faster than it
actually is by virtue of LLVM being extremely slow. The memory locality in
the backend is terrible, for example.

**Comment by thewilsonator** [comment]
This is now green.

**Comment by RazvanN7** [comment]
Since @WalterBright has expressed concern about this PR, I think that ultimately he should be the one to accept this PR.

**Comment by thewilsonator** [comment]
All of his concerns have been addressed. If he has new ones then he has best speak up soon.

**Comment by thewilsonator** [comment]
@RazvanN7 times up. Probably best to squash this.

**Comment by WalterBright** [comment]
Didn't try: "can you try a larger executable?"

**Comment by adamdruppe** [comment]
It is hard to find a large program that actually even works with optlink. The biggest programs I have - two day job applications - just crash the linker.

So gotta grab something medium sized from my other computer and I keep having other things to do. I'll find something today though.

But frankly if the linker crashes for the commercial applications and loses the speed war for simple applications, it isn't a great default regardless of how fast it is for medium sized things.

**Comment by adamdruppe** [comment]
Pain to test with my new game website application since I'd have to convert the database library file format for it.

I know, I can do it, but I mention it because this is a pain new users to D experience pretty frequently too. "just use -m32mscoff lol" is the easier answer than "buy the extended utilities package and run the program"

**Comment by adamdruppe** [comment]
I abandoned optlink entirely on my cross-compilation setup because it would just lock up about 30% of builds forcing me to cancel and retry when it seemed to be taking forever. The -m32mscoff build, which uses lld-link by default in the dmd distribution for quite some time now, just works consistently. Never seen it lock up like that.

**Comment by adamdruppe** [comment]
hah i assumed my web application would be bigger, but it came in at 1.2 MB. I just write overly efficient code apparently.

Nevertheless, its numbers:

MS Link full build: 2.4 s
Optlink full build: 2.4 s
lld-link full build: 2.4s

All within the margin of measurement error.


Well, the nice thing about the web framework is it generates code automatically. Maybe I can just static foreach and add a few thousand functions, then let its templates expand that to more. But the problem is... I write efficient code and avoid unnecessary codegen since I can't stand overly slow compiles. Still worth a try.

Adding 2,000 functions brings it up to 3.6 MB. Still not as big as the day job apps by far, but again, they don't even work at all with optlink alas.

Here we go. Note that linking is going to be a relatively small minority of this time since most of it is spent generating the 2000 functions then generating the 2000 wrappers to present them to the web. Anyway:

optlink full build: 32.4s (so much for my efficient code)
mslink full build: 24.1s ....... ok im surprised, maybe more time spent in the linker than i thought.
lld-link full build: 24.3s

..... optlink got *blown out*. I expected the difference would be within a second or two again. Running it a few more times does show some variation: 31.6 s, 32.8s. MS and lld showed a peak of 27s. We have a margin of error here of 1-3s, but they still consistently win. The slowest MS or LLD link time is still several seconds faster than the fastest optlink time.

But both MS and lld link are well ahead. Doing a `dmd -c` thing just to compare that ... 24-27s. The link time is *negligible* with them,  99% of the time spent in dmd instead of in link (which is exactly what I thought would happen, 2000 functions still isn't a terribly large symbol table), but with optlink we're spending about 15% of the full build time in link alone.

No wonder it has no hope of linking the day job application, which takes 5-10 seconds in the lld-link phase alone with a godawful number of symbols.

I wouldn't be surprised if there's just a linear scan in optlink causing n^2 behavior that could be optimized to a hash table or something to make it competitive again.

But....... who cares? Speed is the least of its problems.

I like optlink. I used it for some seventeen years, the latter ones being when people were pushing lld-link but it didn't just work, so dmd -m32 with optlink was a default I defended since it actually did work out of the box for many things. I still go back to 16 bit DOS code sometimes too where it is fun.

But there's just no reason to keep it the default for dmd anymore. lld-link is packaged in the box and just works both in simple cases and more advanced cases with no need to convert library formats and it consistently does as well or a better job in speed.

**Comment by WalterBright** [comment]
@adamdruppe thanks, this is good information. The two remaining reasons I wanted to keep optlink were having it "in the box", and it being faster. These are no longer the case, so I'm good with leaving it behind.

I could fix it, but the way it is written would consume hundreds of hours of my time (converting it to D). It's not worth it.

It's nice that *finally* other linkers caught up with it. It only took about 30 years :-) That's quite an accomplishment.

**Comment by WalterBright** [comment]
P.S. Nearly all of its bugs appear to be capacity problems. Programs have increased enormously in size.

**Comment by adamdruppe** [comment]
Yeah, the other linkers were *brutally* slow until quite recently, I think it was only about ten years ago that the gold linker became usable and kicked the others back into gear making an effort.

But D programs too tend to generate a large number of symbols. I try to code my things to keep that under control, using techniques like ctfe-only functions or type erasure forwarding. But it can still add up pretty fast.

**Comment by maxhaton** [comment]
See also mold, which is extremely fast (much faster than lld, which in turn
makes gold look like an old man), and increasingly ready for prime time.

lld doesn't support LTO with GCC, which means gold still has to be the
default in a few cases, unfortunately.

I can't find the code but I actually had a kprobe based alarmed that would
go off if it thought GNU LD was linking instead of a fast linker.

**Comment by thewilsonator** [comment]
https://github.com/dlang/druntime/pull/3703

**Comment by schveiguy** [comment]
FYI, there is a potentially significant problem with targeting mscoff by default with -betterC: MS stdio (at least the way we use it) requires runtime hooks to set up `stdout`/`stdin`/`stderr`. See some discussion here: https://forum.dlang.org/post/zqfesvprqujdcvhdedgd@forum.dlang.org

Yes, this is happening before this PR is in the released compiler, but dub has been building using mscoff by default for a while, which could be causing this problem.

I'm not 100% sure, because I don't know what his environment looks like. He says it works on 2.091.0, but the [dub switch to mscoff by default](https://dlang.org/changelog/2.086.0.html#dmd-mscoff-default) was done before that release. Which is why I said "potential". We should at least investigate if it's possible to use libc without druntime for -betterC.

**Comment by strawberry9** [comment]
dmd -m32mscoff -betterC  (no, it does not play well together)

and you can forget about -betterC with -m64 ...  (i.e the stdout/stdin/stderr thing)..

IMO. it's a complete mess on Windows. I don't like it. It's a real turnoff.

I would prefer to have 100% reliance on MSVC. builtools. It's really not that hard to install, and it's not that hard to get the correct ENV settings either.

**Pull request merged**